sync15-adapter/src/request.rs

```diff
@@ -0,0 +1,1203 @@
+    pub successful_ids: Vec<String>,
+    pub allow_failed: bool,
+    pub pending_failed: Vec<String>,
+    pub pending_success: Vec<String>,
+}
+
+impl NormalResponseHandler {
+    pub fn new(allow_failed: bool) -> NormalResponseHandler {
+        NormalResponseHandler {
+            failed_ids: vec![],
+            successful_ids: vec![],
+            pending_failed: vec![],
+            pending_success: vec![],
+            allow_failed,
+        }
+    }
+}
+
+impl PostResponseHandler for NormalResponseHandler {
+    fn handle_response(&mut self, r: PostResponse, mid_batch: bool) -> error::Result<()> {
+        if !r.status.is_success() {
+            warn!("Got failure status from server while posting: {}", r.status);
+            if r.status == StatusCode::PreconditionFailed {
+                bail!(error::ErrorKind::BatchInterrupted);
+            } else {
+                bail!(error::ErrorKind::StorageHttpError(r.status,
+                    "collection storage (TODO: record route somewhere)".into()));
+            }
+        }
+        if r.result.failed.len() > 0 && !self.allow_failed {
+            bail!(error::ErrorKind::RecordUploadFailed(r.result.failed.clone()));
+        }
+        for id in r.result.success.iter() {
+            self.pending_success.push(id.clone());
+        }
+        for kv in r.result.failed.iter() {
+            self.pending_failed.push(kv.0.clone());
+        }
+        if !mid_batch {
+            self.successful_ids.append(&mut self.pending_success);
+            self.failed_ids.append(&mut self.pending_failed);
+        }
+        Ok(())
+    }
+}
+
+impl<Poster, OnResponse> PostQueue<Poster, OnResponse>
+where
+    Poster: BatchPoster,
+    OnResponse: PostResponseHandler
+{
+    pub fn new(config: &InfoConfiguration,
+               ts: ServerTimestamp,
+               poster: Poster,
+               on_response: OnResponse) -> PostQueue<Poster, OnResponse> {
+        PostQueue {
+            poster,
+            on_response,
+            last_modified: ts,
+            post_limits: LimitTracker::new(config.max_post_bytes, config.max_post_records),
+            batch_limits: LimitTracker::new(config.max_total_bytes, config.max_total_records),
+            batch: BatchState::NoBatch,
+            max_payload_bytes: config.max_record_payload_bytes,
+            max_request_bytes: config.max_request_bytes,
+            queued: Vec::new(),
+        }
+    }
+
+    #[inline]
+    fn in_batch(&self) -> bool {
+        match &self.batch {
+            &BatchState::Unsupported |
+            &BatchState::NoBatch => false,
+            _ => true
+        }
+    }
+
+    pub fn enqueue(&mut self, record: &BsoRecord<EncryptedPayload>) -> Result<bool> {
+        let payload_length = record.payload.serialized_len();
+
+        if self.post_limits.can_never_add(payload_length) ||
+           self.batch_limits.can_never_add(payload_length) ||
+           payload_length >= self.max_payload_bytes {
```

Comment:
I wonder if we can just treat this in the same way as the server reporting a failed record?

Reply:
Maybe, I can see more of a case for returning an Err() though. I just adapted the desktop API more or less verbatim though.